core/src/main/scala/sttp/client/testing/SttpBackendStub.scala

```diff
@@ -97,49 +117,84 @@ class SttpBackendStub[F[_], S](
       thenRespondWithCode(StatusCode.Ok)
-    def thenRespondNotFound(): SttpBackendStub[F, S] =
+    def thenRespondNotFound(): SttpBackendStub[F, S, WS_HANDLER] =
       thenRespondWithCode(StatusCode.NotFound, "Not found")
-    def thenRespondServerError(): SttpBackendStub[F, S] =
+    def thenRespondServerError(): SttpBackendStub[F, S, WS_HANDLER] =
       thenRespondWithCode(StatusCode.InternalServerError, "Internal server error")
-    def thenRespondWithCode(status: StatusCode, msg: String = ""): SttpBackendStub[F, S] = {
+    def thenRespondWithCode(status: StatusCode, msg: String = ""): SttpBackendStub[F, S, WS_HANDLER] = {
       thenRespond(Response(msg, status, msg))
     }
-    def thenRespond[T](body: T): SttpBackendStub[F, S] =
+    def thenRespond[T](body: T): SttpBackendStub[F, S, WS_HANDLER] =
       thenRespond(Response[T](body, StatusCode.Ok, "OK"))
-    def thenRespond[T](resp: => Response[T]): SttpBackendStub[F, S] = {
+    def thenRespond[T](resp: => Response[T]): SttpBackendStub[F, S, WS_HANDLER] = {
       val m: PartialFunction[Request[_, _], F[Response[_]]] = {
         case r if p(r) => monad.eval(resp)
       }
-      new SttpBackendStub(monad, matchers.orElse(m), fallback)
+      new SttpBackendStub[F, S, WS_HANDLER](monad, matchers.orElse(m), wsMatchers, fallback)
+    }
+
+    /**
+      * When [[openWebsocket()]] is called, it will ignore handler and return given result.
+      * It is intended to be used when [[openWebsocket()]] is called with sttp supplied handlers
+      * that return wrapped WebSocket as WS_RESULT.
+      * */
+    def thenRespondWebSocket[WS_RESULT](headers: Headers, result: WS_RESULT): SttpBackendStub[F, S, WS_HANDLER] = {
+      val m: PartialFunction[Request[_, _], WhenOpenWebsocket[F, WS_HANDLER]] = {
+        case r if p(r) =>
+          ReturnWebsocketResponse(headers, monad.unit(result))
+      }
+      new SttpBackendStub[F, S, WS_HANDLER](monad, matchers, wsMatchers.orElse(m), fallback)
+    }
+
+    /**
+      * When [[openWebsocket()]] is called, it will ignore handler and return given result.
+      * It is intended to be used when [[openWebsocket()]] is called with sttp supplied handlers
+      * It returns [[WebSocket]] built by [[WebSocketStub]] wrapped as WS_RESULT.
+      * */
+    def thenRespondWebSocket[WS_RESULT](headers: Headers, result: WebSocketStub[_]): SttpBackendStub[F, S, WS_HANDLER] = {
+      thenRespondWebSocket(headers, result.build(monad))
+    }
+
+    /**
+      * When [[openWebsocket()]] is called it uses given headers and handler to create result.
+      * It is intended to be used when [[openWebsocket()]] is called with user supplied handler that
+      * doesn't return WebSocket object to act on.
+      * */
+    def thenHandleOpenWebSocket[WS_RESULT](headers: Headers, useHandler: WS_HANDLER[WS_RESULT] => WS_RESULT) = {
+      val m: PartialFunction[Request[_, _], WhenOpenWebsocket[F, WS_HANDLER]] = {
```

Comment:
would it be possible to somehow expose a variant which would expose the partial function? So that the returned websocket/handler could depend (dynamically) on the request

Reply:
Done